### PR TITLE
Roulette - Controller and spin transaction

### PR DESCRIPTION
**Pull request by joaop21**
This PR contains:
1. PrizeController is the controller that lists and shows prizes.
2. The roulette cost is something environmental, that could vary from an event to event. So this value should be an environment variable loaded in Safira's initialization.
2. Since the attendee's token_balance needs to be updated, a new changeset had to be added.
3. To mock a roulette, there were essentially two ways for generating a random spin, taking into account the probabilities of the prizes. The first was [this](https://dev.to/haile/the-weight-of-choice-is-heavy-55o), and the socond was [this](https://hashrocket.com/blog/posts/the-adventures-of-generating-random-numbers-in-erlang-and-elixir). I chose the second for simplicity of implementation, and because it seemed more intuitive than the first one.
4. To spin the roulette we have to take into account that multiple users can be doing it concurrently, which in a database can mean that we have several transactions decreasing the stock of a given prize. By default postgres uses the Read_Committed isolation level that is not able to solve problems like NonRepeatable_Read or Phantom_Read. For safety reasons, and to make sure that we do not give 2 prizes when the stock is only 1, we had to increase the isolation of transactions to Serializable.

Just a few notes:
1. Prize probabilities are updated when the stock of a prize reaches 0.
2. If there are multiple participants playing the roulette and because we are using the serializable isolation level, the database raises an error that is expected (multiple transactions on the same data). To solve it, it's enough to catch that error and execute the transaction again until it succeeds.
3. The case where the participant wins a prize that has already been awarded the maximum number of times to him is not dealt with. When this happens a message is returned saying that he has already won that prize the maximum number of times possible for an attendee.

Resolves: #142

### CODE REVIEW

File: lib/safira/roulette/roulette.ex [hunk 1]
```diff
@@ -196,4 +205,133 @@ defmodule Safira.Roulette do
...
+      })
+    )
+    # prize after spinning
+    |> Multi.run(:prize, fn _repo, _changes -> {:ok, spin_prize()} end)
+    # get an attendee_prize if exists one, or build one
+    |> Multi.run(:attendee_prize, fn _repo, %{attendee: attendee, prize: prize} ->
+      {:ok,
+       get_keys_attendee_prize(attendee.id, prize.id) ||
+         %AttendeePrize{attendee_id: attendee.id, prize_id: prize.id, quantity: 0}}
+    end)
```
Comment (jpcorreia99) [inline lib/safira/roulette/roulette.ex:230]: Just checking, the `attendee` used at line 226 is the one the `spin` function received by argument  (thus being not the updated version) and not the updated one which resulted from the Multi.Update operation, right?

Comment (jpcorreia99) [reply lib/safira/roulette/roulette.ex:230]: I know this doens't make any difference regarding your code, I was just curious to see if there was any way that elixir/Multi could detect that the attendee had already been updated in a previous operation.

Comment (jpcorreia99) [inline lib/safira/roulette/roulette.ex:230]: If the first Multi.run only yields the repo and the prize, where is the attendee used in line 226 coming from? Does the spin_prize function also return the attendee?

**Review by jpcorreia99 — APPROVED** [review]
This was both a lesson on database isolation, cryptography and how to write amazing code :shipit:

File: lib/safira/roulette/roulette.ex (hunk 1 above)
Comment (joaop21) [reply lib/safira/roulette/roulette.ex:230]: Recall that Multi is a structure that holds the information for the transaction to run. That attendee in line 226 comes from the operation in the line 217. That attendee is the name of that operation. So what i'm doing is accessing the value of that operation that is stored in Multi structure. If the operation was called :op1, the line 226 would be %{op1: attendee,...}.

**Review by JoaoSCosta — APPROVED** [review]
At the moment I'm doing 'serializable transactions' the hard way, but yours looks like a good way to organise it! 😂

**Pull request merged**